File: 227-BasicCalculatorIi/227-BasicCalculatorIi.py
# Last updated: 8/13/2025, 8:21:00 PM
class Solution:
    def calculate(self, s: str) -> int:
        operand = []
        operator = []
        i = 0
        while i < len(s) : 
            if s[i] in ['+' ,'-' , '*' , '/'] : 
                while len(operator) > 0 and self.precedence(operator[-1]) >= self.precedence(s[i]) : 
                    value2 = operand.pop(-1)
                    value1 = operand.pop(-1)
                    op = operator.pop(-1)
                    
                    value = self.calculation(value1, value2 , op)
                    operand.append(value)
                
                operator.append(s[i])
            elif s[i] == '(' : 
                operator.append(s[i])
            elif s[i] == ')' : 
                while operator[-1] != '(' : 
                    value2 = operand.pop(-1)
                    value1 = operand.pop(-1)
                    op = operator.pop(-1)
                    
                    value = self.calculation(value1 , value2 , op)
                    operand.append(value)

                operator.pop(-1)
            elif s[i] != ' ' : 
                number = 0
                while i < len(s) and s[i] not in ['+' ,'-' , '*' , '/' , ' '] : 
                    number = number * 10 + int(s[i])
                    i+=1
                
                operand.append(number)
                i-=1
            i+=1
        
        while len(operator) > 0: 
            value2 = operand.pop(-1)
            value1 = operand.pop(-1)
            op = operator.pop(-1)
                    
            value = self.calculation(value1, value2 , op)
            operand.append(value)
        
        return operand.pop(-1)
        
    
    def calculation(self , a : int , b : int , o : str) -> int : 
        if o == '+' : 
            return a + b
        elif o == '-' : 
            return a - b
        elif o == '*' : 
            return a * b
        else : 
            return a // b

    def precedence(self , c : str) -> int : 
        if c == '*' or c == '/' : 
            return 2
        elif c == '+' or c == '-' : 
            return 1
        return 0

File: 227-BasicCalculatorIi/227-BasicCalculatorIi.java
// Last updated: 8/13/2025, 8:21:08 PM
class Solution {
    public int calculate(String s) {
        Stack<Integer> operand = new Stack();
        Stack<Character> operator = new Stack();
        int i = 0;
        while(i<s.length()){
            char c = s.charAt(i);
            if(c>='0' && c<='9'){
                int num = 0;
                while(i<s.length() && s.charAt(i)>='0' && s.charAt(i)<='9'){
                    num = num *10 + (s.charAt(i) - '0');
                    i++;
                }
                i--;
                operand.add(num);
            }
            else if(c=='+' || c=='-' || c=='/' || c=='*'){
                while(operator.size()>0 && precedence(c)<=precedence(operator.peek())){
                    int val2 = operand.pop();
                    int val1 = operand.pop();
                    char op = operator.pop();
                    int a = calculation(val1,val2,op);
                    operand.push(a);
                }
                operator.push(c);
            }
            i++;
        }

        while(operator.size()>0){
            int val2 = operand.pop();
            int val1 = operand.pop();
            char op = operator.pop();
            int a = calculation(val1,val2,op);
            operand.push(a);
        }
        return operand.pop();
    }

    public int calculation(int a, int b , char c){
        if(c=='+'){
            return a+b;
        }
        else if(c=='-'){
            return a-b;
        }
        else if(c=='*'){
            return a*b;
        }
        return a/b;
    }

    public int precedence(char c){
        if(c=='+' || c=='-'){
            return 0;
        }
        return 1;
    }
}